<a href="https://colab.research.google.com/github/MukulRaj-704/Machine-Learning-Algos/blob/main/Linear%20Regression/Auto_MPG_LR.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

This is a AUTO MPG dataset it is used to preddict the car fuel efficiency (miles per gallons) based on different vehicle characteristics.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:

from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
Auto_MPG=pd.read_csv('/content/drive/My Drive/ML/Datasets/Auto-MPG.csv')

In [4]:
Auto_MPG.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0        130    3504          12.0          70   
1  15.0          8         350.0        165    3693          11.5          70   
2  18.0          8         318.0        150    3436          11.0          70   
3  16.0          8         304.0        150    3433          12.0          70   
4  17.0          8         302.0        140    3449          10.5          70   

   origin                   car name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

df.info() provides a concise summary of a DataFrame, including column names, data types, non-null counts, and memory usage, helping quickly assess data structure and missing values.

In [7]:
Auto_MPG.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    398 non-null    object 
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model year    398 non-null    int64  
 7   origin        398 non-null    int64  
 8   car name      398 non-null    object 
dtypes: float64(3), int64(4), object(2)
memory usage: 28.1+ KB


Calculating Correlation

In [43]:
correlation=Auto_MPG.corr(numeric_only=True)
print(correlation)

                   mpg  cylinders  displacement    weight  acceleration  \
mpg           1.000000  -0.775396     -0.804203 -0.831741      0.420289   
cylinders    -0.775396   1.000000      0.950721  0.896017     -0.505419   
displacement -0.804203   0.950721      1.000000  0.932824     -0.543684   
weight       -0.831741   0.896017      0.932824  1.000000     -0.417457   
acceleration  0.420289  -0.505419     -0.543684 -0.417457      1.000000   
model year    0.579267  -0.348746     -0.370164 -0.306564      0.288137   
origin        0.563450  -0.562543     -0.609409 -0.581024      0.205873   

              model year    origin  
mpg             0.579267  0.563450  
cylinders      -0.348746 -0.562543  
displacement   -0.370164 -0.609409  
weight         -0.306564 -0.581024  
acceleration    0.288137  0.205873  
model year      1.000000  0.180662  
origin          0.180662  1.000000  


From the correlation matrix, cylinders, displacement, and weight are highly correlated, indicating that they contain similar information. Therefore, only one of these features should be selected for model training.

In [10]:
df = Auto_MPG[['mpg','weight','acceleration','model year','origin']]

In [11]:
df.head()

mpg  weight  acceleration  model year  origin
0  18.0    3504          12.0          70       1
1  15.0    3693          11.5          70       1
2  18.0    3436          11.0          70       1
3  16.0    3433          12.0          70       1
4  17.0    3449          10.5          70       1

isnull() checks for missing values in a dataset and returns a Boolean output (True for null, False for non-null), helping identify missing data.

In [14]:
df.isnull().sum()

mpg             0
weight          0
acceleration    0
model year      0
origin          0
dtype: int64

describe() generates descriptive statistics of numerical features, including count, mean, standard deviation, minimum, maximum, and quartiles, helping summarize the data distribution.

In [44]:
df.describe()

mpg       weight  acceleration  model year      origin
count  398.000000   398.000000    398.000000  398.000000  398.000000
mean    23.514573  2970.424623     15.568090   76.010050    1.572864
std      7.815984   846.841774      2.757689    3.697627    0.802055
min      9.000000  1613.000000      8.000000   70.000000    1.000000
25%     17.500000  2223.750000     13.825000   73.000000    1.000000
50%     23.000000  2803.500000     15.500000   76.000000    1.000000
75%     29.000000  3608.000000     17.175000   79.000000    2.000000
max     46.600000  5140.000000     24.800000   82.000000    3.000000

In [45]:
x=df[['weight','acceleration','model year','origin']]
y=df[['mpg']]

In [46]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2, random_state=42)

In [47]:
x_train.head()

weight  acceleration  model year  origin
3      3433          12.0          70       1
18     2130          14.5          70       3
376    2025          18.2          82       3
248    1800          16.4          78       3
177    2694          15.0          75       2

Feature standardization is performed to ensure that all features contribute equally to the training process, preventing dominance by features with larger numerical scales.

Fitting the scaler on test data causes data leakage and leads to overly optimistic model performance. we dont use fit_transform() on test data.

In [48]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train = scaler.fit_transform(x_train)
X_test = scaler.transform(x_test)
X_train

array([[ 0.55282624, -1.31933367, -1.6966673 , -0.72949361],
       [-0.99966729, -0.41318225, -1.6966673 ,  1.73836775],
       [-1.1247723 ,  0.92792185,  1.63897537,  1.73836775],
       ...,
       [-0.54094894, -0.30444408,  0.52709448,  1.73836775],
       [-1.09498539,  0.60170734,  1.36100515,  1.73836775],
       [-1.21413301,  1.94281144, -0.86275663,  0.50443707]])

fit_transform() calculate the mean and sd and transform the data from dataframe to numpy array

transform() used in test data doen not calculate the mean and sd it uses the above calculated mean and sd

In [49]:
from sklearn.linear_model import LinearRegression
LR_model = LinearRegression()
LR_model.fit(X_train, y_train)

LinearRegression()

In [50]:
y_pred=LR_model.predict(X_test)

In [51]:
from sklearn.metrics import r2_score, mean_squared_error
print("R² Score:", r2_score(y_test, y_pred))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred)))

R² Score: 0.8420637095814091
RMSE: 2.9140474729960792


In [52]:
coef = pd.DataFrame({
    'Feature': x_train.columns,
    'Coefficient': LR_model.coef_.flatten()
})
coef

Feature  Coefficient
0        weight    -5.087055
1  acceleration     0.159728
2    model year     2.831225
3        origin     0.881787

R² Score indicate that model explains 84% of the variance in MPG
The RMSE of 2.91 suggests that the model’s predictions deviate from actual MPG values by approximately 3 MPG on average.